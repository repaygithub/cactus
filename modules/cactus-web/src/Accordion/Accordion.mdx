---
name: Accordion
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import Accordion from './Accordion'
import cactusTheme from '@repay/cactus-theme'

# Accordion

## Best practices

- If you need to group accordions together, make sure to place them inside the `Accordion.Provider` component, as it provides the keyboard accessibility features
for groups of accordions.

## Basic usage

You can use a single Accordion by itself:

```jsx
import React from 'react'
import { Accordion, Text } from '@repay/cactus-web'

const SingleAccordion = props => {
  return (
    <Accordion>
      <Accordion.Header>
        <Text as="h3">My Header</Text>
      </Accordion.Header>
      <Accordion.Body>
        Accordion Content Goes Here
      </Accordion.Body>
    </Accordion>
  )
}
```

### Accordion Provider

You can also use `Accordion.Provider` to display multiple Accordions at the same time, and the Provider will manage their open/closed states for you:

When using `Accordion.Provider`, you can set the maximum number of Accordions that can be open at the same time using the `maxOpen` prop, which defaults to 1.

```jsx
import React from 'react'
import { Accordion, Text } from '@repay/cactus-web'

const MultipleAccordions = props => {
  return (
    <Accordion.Provider maxOpen={2}>
      <Accordion>
        <Accordion.Header>
          <Text as="h3">My Header 1</Text>
        </Accordion.Header>
        <Accordion.Body>
          Accordion Content Goes Here
        </Accordion.Body>
      </Accordion>
      <Accordion>
        <Accordion.Header>
          <Text as="h3">My Header 2</Text>
        </Accordion.Header>
        <Accordion.Body>
          Accordion Content Goes Here
        </Accordion.Body>
      </Accordion>
      <Accordion>
        <Accordion.Header>
          <Text as="h3">>My Header 3</Text
        </Accordion.Header>
        <Accordion.Body>
          Accordion Content Goes Here
        </Accordion.Body>
      </Accordion>
    </Accordion.Provider>
  )
}
```

### Outline Variant

We have included an `outline` variant which can be used to better depict hierarchical structures as well as separations in the page.

Usage:

```jsx
import React from 'react'
import { Accordion, Text } from '@repay/cactus-web'

const OutlineAccordion = props => {
  return (
    <Accordion variant="outline">
      <Accordion.Header>
        <Text as="h3">Example Header</Text>
      </Accordion.Header>
      <Accordion.Body>
        Example Content
      </Accordion.Body>
    </Accordion>
  )
}
```

### Render Prop

To allow further customization of the accordions, we've provided the option to pass a render function to the header. This function receives an object
which includes a variable indicating whether or not the accordion is open, as well as an id that can be placed on whichever element describes the accordion.

Usage:

```jsx
import React, { useState } from 'react'
import { Accordion, IconButton, Text } from '@repay/cactus-web'
import { ActionsDelete, NavigationCircleDown, NavigationCircleUp } from '@repay/cactus-icons'

const initialAccordions = [
  {
    header: 'Header 1',
    body: 'Body 1'
  },
  {
    header: 'Header 2'
    body: 'Body 2'
  }
]

const CustomAccordions = props => {
  const [accordions, setAccordions] = useState(initialAccordions)

  const handleUpClick = index => {
    accordionsCopy = [...accordions]
    const temp = accordionsCopy[index - 1]
    accordionsCopy[index - 1] = accordionsCopy[index]
    accordionsCopy[index] = temp
    setAccordions(accordionsCopy)
  }

  const handleDownClick = index => {
    accordionsCopy = [...accordions]
    const temp = accordionsCopy[index + 1]
    accordionsCopy[index + 1] = accordionsCopy[index]
    accordionsCopy[index] = temp
    setAccordions(accordionsCopy)
  }

  const handleDelete = index => {
    const accordionsCopy = [...accordions]
    accordionsCopy.splice(index, 1)
    setAccordions(accordionsCopy)
  }

  return (
    <Accordion.Provider>
      {accordions.map((acc, index) => (
        <Accordion variant="outline" key={acc.header}> render={({ isOpen, headerId }) => (
          <Flex alignItems="center" width="100%>
            <Text as="h3" id={headerId}>{acc.header}</Text>
            {isOpen && (
              <IconButton
                iconSize="medium"
                variant="danger"
                ml="auto"
                mr={4}
                label={`Delete ${acc.header}`}
                onClick={e => {
                  handleDelete(index)
                  e.stopPropagation()
                }}
              >
                <ActionsDelete aria-hidden="true" />
              </IconButton>
            )}
            <Flex
              alignItems="center"
              ml={isOpen ? 0 : 'auto'}
              pl={4}
              borderLeft="1px solid"
              borderLeftColor="lightContrast"
            >
              <IconButton
                iconSize="medium"
                mr={1}
                label={`Move ${acc.header} down`}
                disabled={index === accordions.length - 1}
                onClick={e => {
                  handleDownClick(index)
                  e.stopPropagation()
                }}
              >
                <NavigationCircleDown aria-hidden="true" />
              </IconButton>
              <IconButton
                iconSize="medium"
                label={`Move ${acc.header} up`}
                disabled={index === accordions.length - 1}
                onClick={e => {
                  handleUpClick(index)
                  e.stopPropagation()
                }}
              >
                <NavigationCircleUp aria-hidden="true" />
              </IconButton>
            </Flex>
          </Flex>
        )} />
      ))}
    </Accordion.Provider>
  )
}
```


## Properties

## Accordion
<PropsTable of={Accordion} />

## Accordion Header
<PropsTable of={Accordion} staticProp="AccordionHeader" />

## Accordion Body
<PropsTable of={Accordion} staticProp="AccordionBody" />

## Accordion Provider
<PropsTable of={Accordion} staticProp="AccordionProvider" />
