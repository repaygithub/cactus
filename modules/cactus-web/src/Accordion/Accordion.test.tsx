import { ActionsDelete, NavigationCircleDown, NavigationCircleUp } from '@repay/cactus-icons'
import { fireEvent, screen, waitForElementToBeRemoved } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React, { ReactElement } from 'react'

import animationRender from '../../tests/helpers/animationRender'
import renderWithTheme from '../../tests/helpers/renderWithTheme'
import Flex from '../Flex/Flex'
import KeyCodes from '../helpers/keyCodes'
import IconButton from '../IconButton/IconButton'
import Text from '../Text/Text'
import Accordion from './Accordion'

const openAccordion = async () => animationRender()

const closeAccordion = async (accordionLabel: string) =>
  waitForElementToBeRemoved(screen.getByLabelText(accordionLabel, { selector: '[role="region"]' }))

const getAccordionButton = (accordionLabel: string) =>
  screen.getByLabelText(accordionLabel, { selector: '[data-role="accordion-button"]' })

const ControlledAccordion: React.FC<{
  initOpenId?: string[]
  maxOpen?: number
  useDefaultOpen?: boolean
  changeHook?: (id: string) => void
}> = ({ changeHook, initOpenId = [], maxOpen, useDefaultOpen }) => {
  const [open, setOpen] = React.useState<string[]>(initOpenId)

  const toggleAccordion = (toggleId: string) => {
    changeHook?.(toggleId)
    setOpen((currentOpen) => {
      if (currentOpen.includes(toggleId)) {
        return currentOpen.filter((id) => id !== toggleId)
      }
      return [...currentOpen, toggleId]
    })
  }

  return (
    <Accordion.Provider openId={open} maxOpen={maxOpen} onChange={toggleAccordion}>
      <Accordion id="A1">
        <Accordion.Header>
          <Text as="h3">Accordion 1</Text>
        </Accordion.Header>
        <Accordion.Body>Should show A1</Accordion.Body>
      </Accordion>
      <Accordion defaultOpen={useDefaultOpen} id="A2">
        <Accordion.Header>
          <Text as="h3">Accordion 2</Text>
        </Accordion.Header>
        <Accordion.Body>Should show A2</Accordion.Body>
      </Accordion>
    </Accordion.Provider>
  )
}

const UncontrolledAccordion = () => (
  <Accordion.Provider maxOpen={2}>
    <Accordion>
      <Accordion.Header>
        <Text as="h3">Accordion 1</Text>
      </Accordion.Header>
      <Accordion.Body>A1 Content</Accordion.Body>
    </Accordion>
    <Accordion>
      <Accordion.Header>
        <Text as="h3">Accordion 2</Text>
      </Accordion.Header>
      <Accordion.Body>A2 Content</Accordion.Body>
    </Accordion>
  </Accordion.Provider>
)

describe('component: Accordion', () => {
  describe('Component', () => {
    test('Should render Accordion', () => {
      const { container } = renderWithTheme(
        <Accordion id="accordion">
          <Accordion.Header>
            <Text as="h3">Test Header</Text>
          </Accordion.Header>
          <Accordion.Body>Test Body</Accordion.Body>
        </Accordion>
      )

      expect(container).toHaveTextContent('Test Header')
      expect(container).not.toHaveTextContent('Test Body')
    })

    test('Should render outline accordion', () => {
      const { container } = renderWithTheme(
        <Accordion id="accordion" variant="outline">
          <Accordion.Header>
            <Text as="h3">Test Header</Text>
          </Accordion.Header>
          <Accordion.Body>Test Body</Accordion.Body>
        </Accordion>
      )

      expect(container).toHaveTextContent('Test Header')
      expect(container).not.toHaveTextContent('Test Body')
    })

    test('Should open Accordion when button is clicked', async () => {
      const { container } = renderWithTheme(
        <Accordion>
          <Accordion.Header>
            <Text as="h3">Test Header</Text>
          </Accordion.Header>
          <Accordion.Body>Test Body</Accordion.Body>
        </Accordion>
      )

      const toggleButton = getAccordionButton('Test Header')
      userEvent.click(toggleButton)
      await openAccordion()
      expect(container).toHaveTextContent('Test Body')
    })

    test('Should allow the user to initialize accordions as open', () => {
      const { container } = renderWithTheme(
        <Accordion defaultOpen>
          <Accordion.Header>
            <Text as="h3">Test Header</Text>
          </Accordion.Header>
          <Accordion.Body>Test Body</Accordion.Body>
        </Accordion>
      )

      expect(container).toHaveTextContent('Test Body')
    })

    test('Should support nested accordions', async () => {
      const { container } = renderWithTheme(
        <Accordion>
          <Accordion.Header>
            <Text as="h3">Parent</Text>
          </Accordion.Header>
          <Accordion.Body>
            <Accordion>
              <Accordion.Header>
                <Text as="h4">Child</Text>
              </Accordion.Header>
              <Accordion.Body>Child Content</Accordion.Body>
            </Accordion>
          </Accordion.Body>
        </Accordion>
      )

      const parentButton = getAccordionButton('Parent')
      userEvent.click(parentButton)
      await openAccordion()
      expect(container).toHaveTextContent('Child')

      const childButton = getAccordionButton('Child')
      userEvent.click(childButton)
      await openAccordion()
      expect(container).toHaveTextContent('Child Content')
    })
  })

  describe('Provider', () => {
    test('Should manage Accordion state', async () => {
      const { container } = renderWithTheme(
        <Accordion.Provider>
          <Accordion>
            <Accordion.Header>
              <Text as="h3">Test Header</Text>
            </Accordion.Header>
            <Accordion.Body>Test Body</Accordion.Body>
          </Accordion>
        </Accordion.Provider>
      )

      const toggleButton = getAccordionButton('Test Header')
      userEvent.click(toggleButton)
      await openAccordion()
      expect(container).toHaveTextContent('Test Body')
    })

    test('Should close one Accordion when another opens', async () => {
      const onChange = jest.fn()
      const { container } = renderWithTheme(
        <Accordion.Provider onChange={onChange}>
          <Accordion id="A1">
            <Accordion.Header>
              <Text as="h3">Accordion 1</Text>
            </Accordion.Header>
            <Accordion.Body>Should show first and not second</Accordion.Body>
          </Accordion>
          <Accordion id="A2">
            <Accordion.Header>
              <Text as="h3">Accordion 2</Text>
            </Accordion.Header>
            <Accordion.Body>Should show second and not first</Accordion.Body>
          </Accordion>
        </Accordion.Provider>
      )

      const a1Button = getAccordionButton('Accordion 1')
      const a2Button = getAccordionButton('Accordion 2')
      userEvent.click(a1Button)
      await openAccordion()
      expect(container).toHaveTextContent('Should show first and not second')
      expect(container).not.toHaveTextContent('Should show second and not first')
      expect(onChange).toHaveBeenCalledTimes(1)
      expect(onChange).toHaveBeenLastCalledWith('A1')

      userEvent.click(a2Button)
      await openAccordion()
      await closeAccordion('Accordion 1')
      expect(container).toHaveTextContent('Should show second and not first')
      expect(container).not.toHaveTextContent('Should show first and not second')
      expect(onChange).toHaveBeenCalledTimes(3)
      expect(onChange).toHaveBeenNthCalledWith(2, 'A2')
      expect(onChange).toHaveBeenNthCalledWith(3, 'A1')
    })

    test('Should allow two Accordions to be open at the same time', async () => {
      const onChange = jest.fn()
      const { container } = renderWithTheme(
        <Accordion.Provider maxOpen={2} onChange={onChange}>
          <Accordion id="A1">
            <Accordion.Header>
              <Text as="h3">Accordion 1</Text>
            </Accordion.Header>
            <Accordion.Body>Should show A1</Accordion.Body>
          </Accordion>
          <Accordion id="A2">
            <Accordion.Header>
              <Text as="h3">Accordion 2</Text>
            </Accordion.Header>
            <Accordion.Body>Should show A2</Accordion.Body>
          </Accordion>
        </Accordion.Provider>
      )

      const a1Button = getAccordionButton('Accordion 1')
      const a2Button = getAccordionButton('Accordion 2')
      userEvent.click(a1Button)
      await openAccordion()
      expect(container).toHaveTextContent('Should show A1')
      expect(container).not.toHaveTextContent('Should show A2')
      expect(onChange).toHaveBeenCalledTimes(1)
      expect(onChange).toHaveBeenLastCalledWith('A1')

      userEvent.click(a2Button)
      await openAccordion()
      expect(container).toHaveTextContent('Should show A1')
      expect(container).toHaveTextContent('Should show A2')
      expect(onChange).toHaveBeenCalledTimes(2)
      expect(onChange).toHaveBeenLastCalledWith('A2')
    })

    test('Should allow the user to initialize accordions as open', async () => {
      const onChange = jest.fn()
      const { container } = renderWithTheme(
        <Accordion.Provider maxOpen={2} onChange={onChange}>
          <Accordion defaultOpen>
            <Accordion.Header>
              <Text as="h3">Accordion 1</Text>
            </Accordion.Header>
            <Accordion.Body>Should show A1</Accordion.Body>
          </Accordion>
          <Accordion defaultOpen>
            <Accordion.Header>
              <Text as="h3">Accordion 2</Text>
            </Accordion.Header>
            <Accordion.Body>Should show A2</Accordion.Body>
          </Accordion>
        </Accordion.Provider>
      )

      await openAccordion()
      expect(container).toHaveTextContent('Should show A1')
      expect(container).toHaveTextContent('Should show A2')
      expect(onChange).not.toHaveBeenCalled()
    })

    test('Should allow the user to control which accordions are open', async () => {
      const changeHook = jest.fn()
      const { container } = renderWithTheme(
        <ControlledAccordion initOpenId={['A2']} changeHook={changeHook} />
      )

      expect(container).not.toHaveTextContent('Should show A1')
      expect(container).toHaveTextContent('Should show A2')
      expect(changeHook).not.toHaveBeenCalled()

      userEvent.click(getAccordionButton('Accordion 1'))
      await openAccordion()

      expect(container).toHaveTextContent('Should show A1')
      expect(container).toHaveTextContent('Should show A2')
      expect(changeHook).toHaveBeenCalledTimes(1)
      expect(changeHook).toHaveBeenLastCalledWith('A1')

      userEvent.click(getAccordionButton('Accordion 2'))
      await closeAccordion('Accordion 2')

      expect(container).toHaveTextContent('Should show A1')
      expect(container).not.toHaveTextContent('Should show A2')
      expect(changeHook).toHaveBeenCalledTimes(2)
      expect(changeHook).toHaveBeenLastCalledWith('A2')
    })

    test('Should ignore any non-existent IDs', () => {
      const { container } = renderWithTheme(<ControlledAccordion initOpenId={['foo']} />)

      expect(container).not.toHaveTextContent('Should show A1')
      expect(container).not.toHaveTextContent('Should show A2')
    })

    test('Should ignore maxOpen if the user controls the open accordions', async () => {
      const error = console.error
      console.error = jest.fn()

      const { container } = renderWithTheme(<ControlledAccordion initOpenId={['A1']} maxOpen={1} />)

      expect(container).toHaveTextContent('Should show A1')
      expect(container).not.toHaveTextContent('Should show A2')

      userEvent.click(getAccordionButton('Accordion 2'))
      await openAccordion()

      expect(container).toHaveTextContent('Should show A1')
      expect(container).toHaveTextContent('Should show A2')

      console.error = error
    })

    test('Should ignore defaultOpen props on children if the user controls open accordions', () => {
      const warn = console.warn
      console.warn = jest.fn()

      const { container } = renderWithTheme(
        <ControlledAccordion initOpenId={['A1']} useDefaultOpen={true} />
      )

      expect(container).toHaveTextContent('Should show A1')
      expect(container).not.toHaveTextContent('Should show A2')

      console.warn = warn
    })

    test('Should properly handle Accordion deletion', async () => {
      const TestAccordion = () => {
        const [accordionHeaders, setAccordionHeaders] = React.useState([
          'First Accordion',
          'Second Accordion',
          'Third Accordion',
          'Fourth Accordion',
          'Fifth Accordion',
        ])

        const handleDelete = (index: number) => {
          const headersCopy = [...accordionHeaders]
          headersCopy.splice(index, 1)
          setAccordionHeaders(headersCopy)
        }

        return (
          <Accordion.Provider maxOpen={1}>
            {accordionHeaders.map(
              (header, index): ReactElement => (
                <Accordion variant="outline" key={header}>
                  <Accordion.Header
                    render={({ isOpen, headerId }): ReactElement => {
                      return (
                        <Flex alignItems="center" width="100%">
                          <Text as="h3" id={headerId}>
                            {header}
                          </Text>
                          {isOpen && (
                            <IconButton
                              iconSize="medium"
                              variant="danger"
                              ml="auto"
                              mr={4}
                              label={`Delete ${header}`}
                              onClick={(e: React.MouseEvent<HTMLButtonElement>) => {
                                handleDelete(index)
                                e.stopPropagation()
                              }}
                            >
                              <ActionsDelete aria-hidden="true" />
                            </IconButton>
                          )}
                        </Flex>
                      )
                    }}
                  />
                  <Accordion.Body>This is the body of the {header}</Accordion.Body>
                </Accordion>
              )
            )}
          </Accordion.Provider>
        )
      }

      const { getByLabelText, getByText, queryByText } = renderWithTheme(<TestAccordion />)

      const a1Button = getAccordionButton('First Accordion')
      userEvent.click(a1Button)
      await openAccordion()
      expect(getByText('This is the body of the First Accordion')).toBeVisible()

      const a1DeleteButton = getByLabelText('Delete First Accordion')
      userEvent.click(a1DeleteButton)

      const a2Button = getAccordionButton('Second Accordion')
      userEvent.click(a2Button)
      await openAccordion()
      expect(getByText('This is the body of the Second Accordion')).toBeVisible()

      const a2DeleteButton = getByLabelText('Delete Second Accordion')
      userEvent.click(a2DeleteButton)

      expect(queryByText('First Accordion')).toBeNull()
      expect(queryByText('Second Accordion')).toBeNull()

      expect(getByText('Third Accordion')).toBeVisible()
      expect(queryByText('This is the body of the Third Accordion')).toBeNull()

      expect(getByText('Fourth Accordion')).toBeVisible()
      expect(queryByText('This is the body of the Fourth Accordion')).toBeNull()

      expect(getByText('Fifth Accordion')).toBeVisible()
      expect(queryByText('This is the body of the Fifth Accordion')).toBeNull()
    })
  })

  describe('Keyboard Interactions - Controlled', () => {
    test('DOWN arrow should move focus to the next accordion', () => {
      renderWithTheme(<ControlledAccordion />)

      const a1 = getAccordionButton('Accordion 1')
      fireEvent.focus(a1)
      fireEvent.keyUp(a1, { key: KeyCodes.DOWN })
      expect(document.activeElement).toBe(getAccordionButton('Accordion 2'))
    })

    test('UP arrow should move focus to the previous accordion', () => {
      renderWithTheme(<ControlledAccordion />)

      const a2 = getAccordionButton('Accordion 2')
      fireEvent.focus(a2)
      fireEvent.keyUp(a2, { key: KeyCodes.UP })
      expect(document.activeElement).toBe(getAccordionButton('Accordion 1'))
    })

    test('UP/DOWN arrows should loop on the accordions', () => {
      renderWithTheme(<ControlledAccordion />)

      const a1 = getAccordionButton('Accordion 1')
      const a2 = getAccordionButton('Accordion 2')
      fireEvent.focus(a1)
      fireEvent.keyUp(a1, { key: KeyCodes.UP })
      expect(document.activeElement).toBe(a2)
      fireEvent.keyUp(a2, { key: KeyCodes.DOWN })
      expect(document.activeElement).toBe(a1)
    })

    test('SPACE should open/close the accordion', () => {
      renderWithTheme(<ControlledAccordion />)

      const a = getAccordionButton('Accordion 1')
      fireEvent.focus(a)
      fireEvent.keyUp(a, { key: KeyCodes.SPACE })
      expect(a.getAttribute('aria-expanded')).toBe('true')
      fireEvent.keyUp(a, { key: KeyCodes.SPACE })
      expect(a.getAttribute('aria-expanded')).toBe('false')
    })

    test('RETURN should open/close the accordion', () => {
      renderWithTheme(<ControlledAccordion />)

      const a = getAccordionButton('Accordion 1')
      fireEvent.focus(a)
      fireEvent.keyUp(a, { key: KeyCodes.RETURN })
      expect(a.getAttribute('aria-expanded')).toBe('true')
      fireEvent.keyUp(a, { key: KeyCodes.RETURN })
      expect(a.getAttribute('aria-expanded')).toBe('false')
    })

    test('HOME should focus on the first accordion', () => {
      renderWithTheme(<ControlledAccordion />)

      const a1 = getAccordionButton('Accordion 1')
      const a2 = getAccordionButton('Accordion 2')
      fireEvent.focus(a2)
      fireEvent.keyUp(a2, { key: KeyCodes.HOME })
      expect(document.activeElement).toBe(a1)
    })

    test('END should focus on the last accordion', () => {
      renderWithTheme(<ControlledAccordion />)

      const a1 = getAccordionButton('Accordion 1')
      const a2 = getAccordionButton('Accordion 2')
      fireEvent.focus(a1)
      fireEvent.keyUp(a1, { key: KeyCodes.END })
      expect(document.activeElement).toBe(a2)
    })
  })

  describe('Keyboard Interactions - Uncontrolled', () => {
    test('DOWN arrow should move focus to the next accordion', () => {
      renderWithTheme(<UncontrolledAccordion />)

      const a1 = getAccordionButton('Accordion 1')
      fireEvent.focus(a1)
      fireEvent.keyUp(a1, { key: KeyCodes.DOWN })
      expect(document.activeElement).toBe(getAccordionButton('Accordion 2'))
    })

    test('UP arrow should move focus to the previous accordion', () => {
      renderWithTheme(<UncontrolledAccordion />)

      const a2 = getAccordionButton('Accordion 2')
      fireEvent.focus(a2)
      fireEvent.keyUp(a2, { key: KeyCodes.UP })
      expect(document.activeElement).toBe(getAccordionButton('Accordion 1'))
    })

    test('UP/DOWN arrows should loop on the accordions', () => {
      renderWithTheme(<UncontrolledAccordion />)

      const a1 = getAccordionButton('Accordion 1')
      const a2 = getAccordionButton('Accordion 2')
      fireEvent.focus(a1)
      fireEvent.keyUp(a1, { key: KeyCodes.UP })
      expect(document.activeElement).toBe(a2)
      fireEvent.keyUp(a2, { key: KeyCodes.DOWN })
      expect(document.activeElement).toBe(a1)
    })

    test('SPACE should open/close the accordion', () => {
      renderWithTheme(<UncontrolledAccordion />)

      const a = getAccordionButton('Accordion 1')
      fireEvent.focus(a)
      fireEvent.keyUp(a, { key: KeyCodes.SPACE })
      expect(a.getAttribute('aria-expanded')).toBe('true')
      fireEvent.keyUp(a, { key: KeyCodes.SPACE })
      expect(a.getAttribute('aria-expanded')).toBe('false')
    })

    test('RETURN should open/close the accordion', () => {
      renderWithTheme(<UncontrolledAccordion />)

      const a = getAccordionButton('Accordion 1')
      fireEvent.focus(a)
      fireEvent.keyUp(a, { key: KeyCodes.RETURN })
      expect(a.getAttribute('aria-expanded')).toBe('true')
      fireEvent.keyUp(a, { key: KeyCodes.RETURN })
      expect(a.getAttribute('aria-expanded')).toBe('false')
    })

    test('HOME should focus on the first accordion', () => {
      renderWithTheme(<UncontrolledAccordion />)

      const a1 = getAccordionButton('Accordion 1')
      const a2 = getAccordionButton('Accordion 2')
      fireEvent.focus(a2)
      fireEvent.keyUp(a2, { key: KeyCodes.HOME })
      expect(document.activeElement).toBe(a1)
    })

    test('END should focus on the last accordion', () => {
      renderWithTheme(<UncontrolledAccordion />)

      const a1 = getAccordionButton('Accordion 1')
      const a2 = getAccordionButton('Accordion 2')
      fireEvent.focus(a1)
      fireEvent.keyUp(a1, { key: KeyCodes.END })
      expect(document.activeElement).toBe(a2)
    })
  })

  describe('Render Prop', () => {
    let suppliedHeaderId: string
    const noop = (e: React.MouseEvent<HTMLButtonElement>) => {
      e.stopPropagation()
    }
    test('Should allow developer to customize content in header', async () => {
      const { getByTestId, getByText } = renderWithTheme(
        <Accordion data-testid="accordion" variant="outline">
          <Accordion.Header
            render={({ isOpen, headerId }): React.ReactElement => {
              suppliedHeaderId = headerId
              return (
                <Flex alignItems="center" width="100%">
                  <Text as="h3" id={headerId}>
                    Test Header
                  </Text>
                  {isOpen && (
                    <IconButton
                      iconSize="medium"
                      variant="danger"
                      ml="auto"
                      mr={4}
                      label="Delete"
                      onClick={noop}
                    >
                      <ActionsDelete aria-hidden="true" />
                    </IconButton>
                  )}
                  <Flex
                    alignItems="center"
                    ml={isOpen ? 0 : 'auto'}
                    pl={4}
                    borderLeft="1px solid"
                    borderLeftColor="lightContrast"
                  >
                    <IconButton iconSize="medium" mr={1} label="Move Down" onClick={noop}>
                      <NavigationCircleDown aria-hidden="true" />
                    </IconButton>
                    <IconButton iconSize="medium" label="Move Up" onClick={noop}>
                      <NavigationCircleUp aria-hidden="true" />
                    </IconButton>
                  </Flex>
                </Flex>
              )
            }}
          />
          <Accordion.Body>Test Body</Accordion.Body>
        </Accordion>
      )

      expect(getByText('Test Header').getAttribute('id')).toEqual(suppliedHeaderId)
      const accordion = getByTestId('accordion')
      expect(accordion.querySelector('button[aria-label="Move Down"]')).toBeInTheDocument()
      expect(accordion.querySelector('button[aria-label="Move Up"]')).toBeInTheDocument()
      expect(accordion.querySelector('button[aria-label=Delete]')).not.toBeInTheDocument()

      const accordionButton = accordion.querySelector(
        'button[data-role="accordion-button"]'
      ) as HTMLButtonElement
      userEvent.click(accordionButton)
      await openAccordion()

      expect(accordion.querySelector('button[aria-label=Delete]')).toBeInTheDocument()
    })

    test('Clicking icon buttons inside header should not trigger opening/closing', async () => {
      const noop = jest.fn()
      const { getByTestId, container } = renderWithTheme(
        <Accordion>
          <Accordion.Header
            render={(): React.ReactElement => (
              <Flex alignItems="center" width="100%">
                <Text as="h3">Test Header</Text>
                <IconButton
                  iconSize="medium"
                  variant="danger"
                  ml="auto"
                  label="Delete"
                  onClick={noop}
                  data-testid="delete"
                >
                  <ActionsDelete aria-hidden="true" />
                </IconButton>
              </Flex>
            )}
          />
          <Accordion.Body>test</Accordion.Body>
        </Accordion>
      )

      const deleteButton = getByTestId('delete')
      userEvent.click(deleteButton)
      await openAccordion()
      expect(noop).toHaveBeenCalled()
      expect(container).not.toHaveTextContent('Should not show')
    })
  })

  describe('With theme customization', () => {
    test('Simple accordion should have 2px border', () => {
      const { container } = renderWithTheme(
        <Accordion id="accordion">
          <Accordion.Header>
            <Text as="h3">Test Header</Text>
          </Accordion.Header>
          <Accordion.Body>Test Body</Accordion.Body>
        </Accordion>,
        { border: 'thick' }
      )

      const accordion = container.querySelector('#accordion')
      const styles = window.getComputedStyle(accordion as Element)
      expect(styles.borderBottomWidth).toBe('2px')
    })

    test('Outline accordion should have 4px border-radius', () => {
      const { container } = renderWithTheme(
        <Accordion id="accordion" variant="outline">
          <Accordion.Header>
            <Text as="h3">Test Header</Text>
          </Accordion.Header>
          <Accordion.Body>Test Body</Accordion.Body>
        </Accordion>,
        { shape: 'intermediate' }
      )

      const accordion = container.querySelector('#accordion')
      const styles = window.getComputedStyle(accordion as Element)
      expect(styles.borderRadius).toBe('4px')
    })

    test('Outline accordion should have 0px border-radius', () => {
      const { container } = renderWithTheme(
        <Accordion id="accordion" variant="outline">
          <Accordion.Header>
            <Text as="h3">Test Header</Text>
          </Accordion.Header>
          <Accordion.Body>Test Body</Accordion.Body>
        </Accordion>,
        { shape: 'square' }
      )

      const accordion = container.querySelector('#accordion')
      const styles = window.getComputedStyle(accordion as Element)
      expect(styles.borderRadius).toBe('0px')
    })
  })

  test('Should support flex item props', () => {
    const { getByTestId } = renderWithTheme(
      <Accordion data-testid="flex-accordion" flex={1} flexGrow={1} flexShrink={0} flexBasis={0}>
        <Accordion.Header>
          <Text as="h3">I Have Flex Item Props</Text>
        </Accordion.Header>
        <Accordion.Body>La dee da</Accordion.Body>
      </Accordion>
    )

    const accordion = getByTestId('flex-accordion')
    expect(accordion).toHaveStyle('flex: 1')
    expect(accordion).toHaveStyle('flex-grow: 1')
    expect(accordion).toHaveStyle('flex-shrink: 0')
    expect(accordion).toHaveStyle('flex-basis: 0')
  })
})
