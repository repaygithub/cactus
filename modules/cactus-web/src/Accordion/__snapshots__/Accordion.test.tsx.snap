// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: Accordion Component Should render Accordion 1`] = `
.c3 {
  vertical-align: middle;
}

.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 1px;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
  color: hsl(200,96%,11%);
  margin-right: 16px;
  font-size: 16px;
}

.c2::-moz-focus-inner {
  border: 0;
}

.c2:focus {
  background-color: hsla(200,96%,35%,0.25);
}

.c2:hover {
  color: hsl(200,96%,35%);
}

.c1 {
  box-sizing: border-box;
  width: 100%;
  min-height: 48px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 600;
  cursor: pointer;
  background: none;
  border: 2px transparent;
  outline: none;
}

.c1::-moz-focus-inner {
  border: 0;
}

.c1 p,
.c1 h1,
.c1 h2,
.c1 h3,
.c1 h4 {
  margin: 0;
}

.c0 {
  width: 100%;
  border-bottom: 2px solid hsl(200,29%,90%);
}

.c0:first-of-type {
  border-top: 2px solid hsl(200,29%,90%);
}

<div>
  <div
    class="c0"
    id="accordion"
  >
    <div
      class="c1"
      id="accordion-header"
    >
      <button
        aria-controls="accordion-body"
        aria-expanded="false"
        aria-labelledby="accordion-header"
        class="c2"
        data-role="accordion-button"
        role="button"
        type="button"
        variant="standard"
      >
        <svg
          aria-hidden="true"
          class="c3"
          fill="currentcolor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M17.7071068,2.70710678 C18.0976311,2.31658249 18.0976311,1.68341751 17.7071068,1.29289322 C17.3165825,0.902368927 16.6834175,0.902368927 16.2928932,1.29289322 L6.29289322,11.2928932 C5.90236893,11.6834175 5.90236893,12.3165825 6.29289322,12.7071068 L16.2928932,22.7071068 C16.6834175,23.0976311 17.3165825,23.0976311 17.7071068,22.7071068 C18.0976311,22.3165825 18.0976311,21.6834175 17.7071068,21.2928932 L8.41421356,12 L17.7071068,2.70710678 Z"
            transform="matrix(-1 0 0 1 24 0)"
          />
        </svg>
      </button>
      <h3
        class=""
      >
        Test Header
      </h3>
    </div>
  </div>
</div>
`;
